buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java

```diff
@@ -1373,12 +1377,17 @@ protected final void checkIndex(int index, int fieldLength) {
     }
 
     final void checkIndex0(int index, int fieldLength) {
-        if (isOutOfBounds(index, fieldLength, capacity())) {
-            throw new IndexOutOfBoundsException(String.format(
-                    "index: %d, length: %d (expected: range(0, %d))", index, fieldLength, capacity()));
+        int capacity = capacity();
```

Comment:
Yes, it is, capacity() is a virtual call which should only be called once.
Also, it helps JVM inlining to move the message creation code into a separate method.